### ISSUE TITLE
Markdown parser gets multiple-underscores-inside-italics wrong

### ISSUE THREAD

**Issue by ljw1004** [tpetricek/FSharp.Formatting#389]
The markdown parser gets underscores-inside-italics wrong. For instance,
```
A *asterisk_one_two* B
```

On both GitHub and CommonMark, it's rendered as an italic with underscores:

A *asterisk_one_two* B

But FSharp.Formatting.Markdown parses it as a nested emphasis:
```
Emphasis [ Literal "asterisk", Emphasis [ Literal "one"], Literal "two"]
```

(Actually, to be honest, I don't know what is the correct behavior. I'm assuming that CommonMark is correct here.)

Here are some more examples:
```
A *asterisk_one_two* B

A *asterisk_one_two_three* B

A *asterisk_one_two_three_four* B

A _underscore_one_two_ B

A _underscore_one_two_three_ B

A _underscore_one_two_three_four_ B
```

A *asterisk_one_two* B

A *asterisk_one_two_three* B

A *asterisk_one_two_three_four* B

A _underscore_one_two_ B

A _underscore_one_two_three_ B

A _underscore_one_two_three_four_ B

**Comment by ljw1004** [comment]
I wonder if this is the same as #376 ?

**Comment by matthid** [comment]
thanks for reporting, yes we have tests for the full spec (Im pretty sure for both issues you reported). If you need a near/mid term fix feel free to enable and fix them. There is not much that can go wrong.

**Comment by matthid** [comment]
If you need any help/pointers feel free to ask.

**Comment by tpetricek** [comment]
I think the behaviour was a bit unspecified before CommonMark.

With CommonMark, I think the rules are that there should be spaces if you want nesting:

    this *is_just_underscores*
    this *is _nested_ emphasis*

**Comment by ljw1004** [comment]
This might be related to #376

**Comment by tpetricek** [comment]
Yes, it is. In the previous issue, @amazingant points to the relevant bits of the CommonMark spec:

> I can't find a part of the CommonMark spec that explicitly states it rather than just being implied by the [left/right flanking delimiter][1] rules, but emphasis with the underscore like `some_words_here` should be translated as `some_words_here`, but come out of FSharp.Markdown as `some<em>words</em>here`. Spec has a [few examples of this][2]; makes snake-case value names output wrong :grimacing:

[1]: http://spec.commonmark.org/0.23/#left-flanking-delimiter-run
[2]: http://spec.commonmark.org/0.23/#example-326

**Comment by deviousasti** [comment]
Has there been any resolution to this?
Working around this is a bit ugly.

Does `DelimitedMarkdown` need to be rewritten?
It looks like the spec has a lot of corner cases.

**Comment by abelbraaksma** [comment]
This is a rather unfortunate bug and as issue #2411 shows, leads to "real world" issues. In that particular scenario for Fantomas, the headers are the names of configuration tokens, which all have underscores, making the file as a whole hard to read.

I don't mind taking a stab at this sometime soon (no promises), perhaps @dsyme or @tpetricek know off-hand where things like this get parsed/rendered? Seems to me that the parser needs some look-ahead/behind to determine that `_` has spaces around.

**Comment by nhirschey** [comment]
I believe it’s here

https://github.com/fsprojects/FSharp.Formatting/blob/a55c29b2ed17b1dc5c98f5bd51e007c86bee82d7/src/FSharp.Formatting.Markdown/MarkdownParser.fs#L165-L181

**Comment by abelbraaksma** [comment]
Thanks for fixing this @nojaf and fr merging it in @dsyme!
